src/NuGet.Core/NuGet.Packaging/Rules/DependeciesGroupsForEachTFMRule.cs

```diff
@@ -0,0 +1,190 @@
+        {
+            var noExactMatchString = new StringBuilder();
+            var compatMatchString = new StringBuilder();
+
+            var beginning = AnalysisResources.DependenciesGroupsForEachTFMBeginningToNuspec;
+            var ending = " " + AnalysisResources.DependenciesGroupsForEachTFMEndingToNuspec;
+
+            var firstElement = new StringBuilder();
+
+            if (noExactMatchesFromFile.Count != 0)
+            {
+                firstElement.Append(beginning);
+                firstElement.Append(noExactMatchesFromFile.Select(t => t.GetFrameworkString()).First());
+
+                noExactMatchString.Append(noExactMatchesFromFile.Count > 1 ? beginning +
+                    string.Join(ending + beginning,
+                        noExactMatchesFromFile.Select(t => t.GetFrameworkString()).ToArray()) :
+                        firstElement.ToString());
+                noExactMatchString.Append(ending);
+            }
+
+            if (noExactMatchesFromNuspec.Count != 0)
+            {
+                beginning = AnalysisResources.DependenciesGroupsForEachTFMBeginningToFiles;
+                ending = " " + AnalysisResources.DependenciesGroupsForEachTFMEndingToFile;
+                firstElement.Clear();
+                firstElement.Append(beginning);
+                firstElement.Append(noExactMatchesFromNuspec.Select(t => t.GetShortFolderName()).First());
+
+                noExactMatchString.Append(noExactMatchesFromNuspec.Count > 1 ? beginning +
+                    string.Join(ending + beginning,
+                        noExactMatchesFromNuspec.Select(t => t.GetShortFolderName()).ToArray()) :
+                    firstElement.ToString());
+                noExactMatchString.Append(ending);
+            }
+
+            if (compatNotExactMatches.Count != 0)
+            {
+                beginning = AnalysisResources.DependenciesGroupsForEachTFMBeginningToNuspec;
+                ending = " " + AnalysisResources.DependenciesGroupsForEachTFMEndingToNuspec;
+                firstElement.Clear();
+                firstElement.Append(beginning);
+                firstElement.Append(compatNotExactMatches.Select(t => t.GetFrameworkString()).First());
+
+                compatMatchString.Append(compatNotExactMatches.Count > 1 ? beginning +
+                    string.Join(ending + beginning,
+                        compatNotExactMatches.Select(t => t.GetFrameworkString()).ToArray()):
+                    firstElement.ToString());
+                compatMatchString.Append(ending);
+            }
+
+            return (noExactMatchString.ToString(), compatMatchString.ToString());
+        }
+
+        private static IEnumerable<NuGetFramework> ExtractTFMsFromNuspec(Stream packageNuspec)
+        {
+            var tfmsFromNuspec = new List<NuGetFramework>();
+            var stream = new NuspecReader(packageNuspec);
```

Comment:
I think the variable names got mixed up. The `Stream`'s name is `packageNuspec`, and the `NuspecReader`'s name is `stream`, which seems backwards.